src/apps/lwaftr/ndp.lua

```diff
@@ -0,0 +1,373 @@
+ipv6_solicited_multicast = ipv6:pton("ff02:0000:0000:0000:0000:0001:ff00:00")
+
+
+-- Given a pointer to 40 bytes representing an IPv6 header,
+-- calculate the checksum the pseudo-header would have if
+-- the pseudo-header werer actually constructed.
+local _scratch_pseudoheader = ffi.new('uint8_t[?]', ipv6_pseudoheader_size)
+local function checksum_pseudoheader_from_header(ipv6_fixed_header)
+   local ph_size = ipv6_pseudoheader_size
+   ffi.fill(_scratch_pseudoheader, ph_size)
+   ffi.copy(_scratch_pseudoheader, ipv6_fixed_header + o_ipv6_src_addr, 32)
+   ffi.copy(_scratch_pseudoheader + 34, ipv6_fixed_header + o_ipv6_payload_len, 2)
+   ffi.copy(_scratch_pseudoheader + 39, ipv6_fixed_header + o_ipv6_next_header, 1)
+   return checksum.ipsum(_scratch_pseudoheader, ph_size, 0)
+end
+
+local function eth_next_is_ipv6(pkt)
+   return rd16(pkt.data + o_ethernet_ethertype) == n_ethertype_ipv6
+end
+
+local function ipv6_next_is_icmp6(pkt)
+   local offset = ethernet_header_size + o_ipv6_next_header
+   return pkt.data[offset] == proto_icmpv6
+end
+
+local function icmpv6_type_is_na(pkt)
+   return pkt.data[eth_ipv6_size] == icmpv6_na
+end
+
+local function icmpv6_type_is_ns(pkt)
+   return pkt.data[eth_ipv6_size] == icmpv6_ns
+end
+
+-- The relevant byte is:
+-- [ R S O 0 0 0 0 0 0 ], where S = solicited
+local function is_solicited_na(pkt)
+   local offset = eth_ipv6_size + 4
+   return bit.band(0x40, pkt.data[offset]) == 0x40
+end
+
+function is_ndp(pkt)
+   if pkt.length > 57 and
+      eth_next_is_ipv6(pkt) and
+      ipv6_next_is_icmp6(pkt)
+   then
+      local icmp_type = pkt.data[eth_ipv6_size]
+      return (icmp_type >= 133 and icmp_type <= 137)
+   end
+   return false
+end
+
+-- TODO: tune this for speed
+-- Note that this must be run after any vlan tags have been stripped out
+-- and after any reassembly has been done, and no extra IPv6 headers
+-- must be present
+--- TODO: could this reasonably use pflang?
+function is_solicited_neighbor_advertisement(pkt)
+   return is_ndp(pkt) and
+          icmpv6_type_is_na(pkt) and
+          is_solicited_na(pkt)
+end
+
+local function is_neighbor_solicitation(pkt)
+   return is_ndp(pkt) and icmpv6_type_is_ns(pkt)
+end
+
+-- Check to see whether the neighbor solicitation is for
+-- one of the listed addresses, rather than an arbitrary one.
+function is_neighbor_solicitation_for_ips(pkt, local_ips)
+   if not is_neighbor_solicitation(pkt) then return false end
+   -- The 'target address' field
+   local target_offset = eth_ipv6_size + 8
+   for i=1,#local_ips do
+      if C.memcmp(local_ips[i], pkt.data + target_offset, 16) == 0 then
+         return true
+       end
+   end
+   return false
+end
+
+function get_dst_ethernet(pkt, target_ipv6_addrs)
+   if pkt == nil or target_ipv6_addrs == nil then return false end
+   local na_addr_offset = eth_ipv6_size + 8
+   for i=1,#target_ipv6_addrs do
+      if C.memcmp(target_ipv6_addrs[i], pkt.data + na_addr_offset, 16) == 0 then
```

Comment:
I think you want ipv6_equals() from lwaftr.lua.  I have to say also that we should be using more pointers into packets like the lwaftr does -- e.g. a pointer to the start of the ND payload start.  However I don't think that should block this merge.